### PR TITLE
Edit feature added to custom toolbar button

### LINKED ISSUE #10436: [Feature request] Edit option on custom toolbar button dialog
**Is your feature request related to a problem? Please describe.**
When long tapping a custom button added to the note editing toolbar, the only option shown is to delete the button. If someone wants to edit the letter or the html tags, their only option is to delete it and create another

**Describe the solution you'd like**
Add an option to edit the button on the dialog

**Additional context**
Current behavior

Screenshot_20220305-173934_AnkiDroid

### PR DESCRIPTION
**Pull request by BishyanKar**
## Purpose / Description
Adding edit feature to custom toolbar button in Add Note screen.

## Fixes
Fixes #10436

## Approach
- Added "Edit" button to the remove dialog and changed its text slightly to suggest the feature. 
- On clicking "Edit" button a dialog similar to create item appears with "Create" button replaced with "Save"
- On clicking "Save" saved the changes to the button

## How Has This Been Tested?

Tested it manually.

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

feature_edit_option_6
feature_edit_option_5

https://user-images.githubusercontent.com/54733471/157710110-2563733b-9cdf-49f4-a7cc-ef19bb61ebff.mp4

### CODE REVIEW

File: AnkiDroid/src/main/res/values/02-strings.xml [hunk 1]
```diff
@@ -355,8 +355,10 @@
     <string name="before_text">HTML Before Selection</string>
     <string name="after_text">HTML After Selection</string>
     <string name="add_toolbar_item">Create Toolbar Item</string>
+    <string name="edit_toolbar_item">Edit Toolbar Item</string>
     <string name="toolbar_item_explain_remove">Enter HTML to be inserted before and after the selected text\n\nLong press a toolbar item to remove it</string>
     <string name="remove_toolbar_item">Remove Toolbar Item?</string>
+    <string name="remove_or_edit_toolbar_item">Remove or Edit Toolbar Item?</string>
```
Comment (david-allison) [inline AnkiDroid/src/main/res/values/02-strings.xml:361]:
```suggestion
    <string name="remove_or_edit_toolbar_item">Edit or Delete Toolbar Item?</string>
```

Comment (david-allison) [reply AnkiDroid/src/main/res/values/02-strings.xml:361]: Keep the text congruent with the button options.

**Review by david-allison — CHANGES_REQUESTED** [review]
Looks great! A few lint errors were tripped, and a couple questions

`gradlew lint` should pick these up so you don't need to use CI to test

```
> Task :AnkiDroid:lintPlayDebug FAILED
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/02-strings.xml:360: Error: The resource R.string.remove_toolbar_item appears to be unused [UnusedResources]
    <string name="remove_toolbar_item">Remove Toolbar Item?</string>
            ~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/02-strings.xml:104: Error: Duplicate string value "Save", used in dialog_positive_save, save [DuplicateCrowdInStrings from com.ichi2.anki.lint]
    <string name="save">Save</string>
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/03-dialogs.xml:116: Duplicates value in save. Add a comment attribute on both strings to explain this duplication
2 errors, 0 warnings
```

To fix:

* Delete `remove_toolbar_item` from `/values`
* Remove `dialog_positive_save` and use `R.string.save`

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java [hunk 2]
```diff
@@ -1985,20 +1987,40 @@ private void removeButton(CustomToolbarButton button) {
...
+        getToolbarDialogBuilder()
+                .title(R.string.edit_toolbar_item)
+                .positiveText(R.string.dialog_positive_save)
                 .onPositive((m, v) -> {
                     View view = m.getView();
                     EditText etIcon =  view.findViewById(R.id.note_editor_toolbar_item_icon);
                     EditText et =  view.findViewById(R.id.note_editor_toolbar_before);
                     EditText et2 = view.findViewById(R.id.note_editor_toolbar_after);
 
+                    removeButton(currentButton);
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java:2023]: Can we fix this to ensure the buttons stay in order?

**Comment by BrayanDSO** [comment]
My personal opinion:
When I use something like this in any app, the default dialog is the edit dialog, and deleting is just a option there.

This is the contrary. The default dialog is the delete and editing is a option. Having to tap an additional button to edit, which is a more common task than deleting, adds friction to the user.

I suggest opening the edit dialog by default, with a delete button below, which leaves less friction.
But, as I said in the beginning, this is just my opinion

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (BishyanKar) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: Done

File: AnkiDroid/src/main/res/values/02-strings.xml (hunk 1 above)
Comment (BishyanKar) [reply AnkiDroid/src/main/res/values/02-strings.xml]: Done

**Comment by david-allison** [comment]
> My personal opinion: When I use something like this in any app, the default dialog is the edit dialog, and deleting is just a option there.

I was considering asking for this, and agree

**Comment by BishyanKar** [comment]
@BrayanDSO @david-allison Alright, I will try and get this done asap.

**Comment by BrayanDSO** [comment]
And I think the edit dialog should be populated with the previous values

**Comment by BishyanKar** [comment]
@BrayanDSO Yes I had this in mind as well. Added it in this commit.

**Comment by Akshay0701** [comment]
Thanks for your contributions! 

apart from code, I have suggestions for UI, 

it's an inconsistent UI, instead of adding the button "delete" at the bottom of the view, why don't you add delete icon button at top right corner of the view, it's just a suggestion and open to other's feedback :) 

Capture

**Comment by BrayanDSO** [comment]
> Keep the text congruent with the button options.
👆

> it's an inconsistent UI, instead of adding the button "delete" at the bottom of the view, why don't you add delete icon button at top right corner of the view, it's just a suggestion and open to other's feedback :)

I agree. Maybe reuse the `Delete note` icon or an X in the top right

**Comment by BrayanDSO** [comment]
> Keep the text congruent with the button options.
👆

> it's an inconsistent UI, instead of adding the button "delete" at the bottom of the view, why don't you add delete icon button at top right corner of the view, it's just a suggestion and open to other's feedback :)

I agree. Maybe reuse the `Delete note` icon or an X in the top right

**Comment by BishyanKar** [comment]
WhatsApp Image 2022-03-13 at 12 13 29
WhatsApp Image 2022-03-13 at 12 13 29 (1)

@BrayanDSO @Akshay0701 which one looks better? Or should I try something else?

**Comment by BrayanDSO** [comment]
> WhatsApp Image 2022-03-13 at 12 13 29 (1)

Pretty neat. 
I personally prefer the trashbin icon. I think it would be better if a little smaller and with one of the theme colors (the same of the buttons would be best I guess).

**Comment by david-allison** [comment]
I don't see the icon in the diff, but I'd also prefer if it wasn't red.

**Comment by BishyanKar** [comment]
WhatsApp Image 2022-03-13 at 20 23 38
@BrayanDSO

**Comment by BrayanDSO** [comment]
> WhatsApp Image 2022-03-13 at 20 23 38

Wonderful! For me, this is it. 

Let's hear the others.

A night mode screenshot would be nice too

**Comment by david-allison** [comment]
Looks good! Could you push the code so we can check it out on our devices?

**Comment by BishyanKar** [comment]
> Looks good! Could you push the code so we can check it out on our devices?

Pushed it.

**Comment by BishyanKar** [comment]
> I like this in general, thank you very much for posting it! However it appears that in your version control you have done a merge commit instead of a rebase commit, so I can't tell exactly what changed. Briefly change to my perspective as a reviewer trying to make sure the source code edits are good and imagine reviewing this - https://github.com/ankidroid/Anki-Android/pull/10464/files 😅
> 
> It's not possible. Can you do a _rebase_ merge from main?

Noticed it, sorry for the mistake. Trying to fix it.

**Comment by david-allison** [comment]
Something's gone wrong with the merge, please fix it, ideally by `rebasing` out the bad commits, or `rebasing` this branch onto the current `main`

**Comment by BishyanKar** [comment]
Couldn't find a way around for the version control issue, it's getting worse, so I had to create a separate pull request. Sorry for the inconvenience caused and would like request you guys to review the other PR. (#10594).

**Comment by david-allison** [comment]
> Couldn't find a way around for the version control issue, it's getting worse, so I had to create a separate pull request. Sorry for the inconvenience caused and would like request you guys to review the other PR. (#10594).

Please spend more time trying to rebase this branch's commits onto `main`. Otherwise we lose the review history.

**Review by mikehardy — APPROVED** [review]
Wow, this is a serious feature! LGTM, requesting a quick look from others for final go

**Comment by BishyanKar** [comment]
@david-allison fixed the version control issue.

**Review by david-allison — APPROVED** [review]
Looks good to me. One copyright question

File: AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml [hunk 3]
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
```
[the remark this answers is not in the document]
Comment (BishyanKar) [reply AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml]: @david-allison Correct, changed it.

File: AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            android:textColor="?attr/colorAccent"
+            android:textSize="20sp"
+            android:textStyle="bold"
+            android:paddingBottom="15dp"
+            app:layout_constraintTop_toTopOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            />
+        <ImageButton
+            android:id="@+id/note_editor_toolbar_btn_delete"
+            android:layout_width="25dp"
```
Comment (david-allison) [inline AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml:44]: Touch targets should be 48x48 for accessibility, and should have a description so blind users know what the icon does.

The icon doesn't need to be 48x48, but the padding/margin should add up to 48x48 to allow users to click.

This is handled by the item in the PR description: 'UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)'

**Review by david-allison — CHANGES_REQUESTED** [review]
Caught an accessibility issue

File: AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml (hunk 4 above)
Comment (BishyanKar) [reply AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml]: 76f57ac4-b66e-4e12-85b8-8053edcf3ee1
1c5f3c63-9185-4cea-a0ec-38e63d6f8b8a


@david-allison modified the xml file accordingly. Attached ss above.

**Review by david-allison — APPROVED** [review]
Thanks!!

File: AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/res/layout/note_editor_toolbar_edit_custom_item.xml]: Fantastic catch! We have not historically done a good job with accessibility (I personally have done a terrible job, I'm not casting blame) so at least myself, I missed that. Great you caught it David

**Comment by mikehardy** [comment]
Hi there! APRIL OpenCollective Notice

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for [April 2022 submissions](https://github.com/ankidroid/Anki-Android/pulls?q=is%3Apr+merged%3A2022-04-01..2022-04-30)

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of March.

Please note that GSoC contributions are okay for this process. Our philosophy is that our users have donated to AnkiDroid for all contributions. The only PRs that will not go through the OpenCollective process are ones directly related to am accepted GSoC project from a selected participant, since those receive a stipend from GSoC itself.

Please understand that this means the monthly budget we have for payments might be stretched a bit and the cap on payments-per-person may be lower, but we try to make our process as fair and transparent as possible, we just need your understanding.

Thanks!

**Pull request merged**